projects/charger/src/generic_can_network.c

```diff
@@ -0,0 +1,95 @@
+#include "generic_can_network.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "can.h"
+#include "generic_can.h"
+#include "generic_can_helpers.h"
+#include "generic_can_msg.h"
+#include "log.h"
+#include "status.h"
+
+static GenericCanInterface s_interface;
+
+// CanRxHandlerCb
+static StatusCode prv_generic_can_network_rx_handler(const CANMessage *msg, void *context,
+                                                     CANAckStatus *ack_reply) {
+  (void)ack_reply;
+  GenericCanRxStorage *rxs = context;
+  if (!rxs->enabled) {
+    return STATUS_CODE_OK;
+  }
+  GenericCanMsg generic_msg = { 0 };
+  can_message_to_generic_can_message(msg, &generic_msg);
+
+  rxs->rx_handler(&generic_msg, rxs->context);
+
+  return STATUS_CODE_OK;
+}
+
+// tx
+static StatusCode prv_tx(const GenericCan *can, const GenericCanMsg *msg) {
+  GenericCanNetwork *gcn = (GenericCanNetwork *)can;
+  if (gcn->base.interface != &s_interface) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "GenericCan not aligned to GenericCanNetwork.");
+  }
+  CANMessage can_msg = { 0 };
+  status_ok_or_return(generic_can_msg_to_can_message(msg, &can_msg));
+
+  return can_transmit(&can_msg, NULL);
+}
+
+// register_rx
+static StatusCode prv_register_rx(GenericCan *can, GenericCanRx rx_handler, uint32_t id,
+                                  void *context) {
+  GenericCanNetwork *gcn = (GenericCanNetwork *)can;
+  if (gcn->base.interface != &s_interface) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "GenericCan not aligned to GenericCanNetwork.");
+  }
+
+  for (size_t i = 0; i < NUM_GENERIC_CAN_NETWORK_RX_HANDLERS; i++) {
+    if (gcn->base.rx_storage[i].rx_handler == NULL && !gcn->base.rx_storage[i].enabled) {
+      gcn->base.rx_storage[i].id = id;
+      gcn->base.rx_storage[i].rx_handler = rx_handler;
+      gcn->base.rx_storage[i].context = context;
+      gcn->base.rx_storage[i].enabled = true;
+      status_ok_or_return(can_register_rx_handler(id, prv_generic_can_network_rx_handler,
```

Comment:
This id will be the message ID, not raw ID.

Reply:
Good idea. Done!